quantization/language_model/llama/weight_only_quant/main.py

```diff
@@ -140,79 +144,26 @@ def eval_func(model):
-        return (torch.vstack(input_ids_padded), torch.vstack(attention_mask_padded)), torch.tensor(last_ind)
-
-
-    def __iter__(self):
-        try:
-            for (input_ids, attention_mask), last_ind in self.dataloader:
-                if self.sess is None:
-                    yield {"input_ids": input_ids[:, :-1].detach().cpu().numpy().astype("int64"),
-                           "attention_mask":attention_mask[:, :-1].detach().cpu().numpy().astype("int64")}, last_ind.detach().cpu().numpy()
-                else:
-                    outputs = self.sess.run(None, {"input_ids": input_ids[:, :-1].detach().cpu().numpy().astype("int64"),
-                                                   "attention_mask":attention_mask[:, :-1].detach().cpu().numpy().astype("int64")})
-                    ort_input = {}
-                    ort_input["input_ids"] = input_ids[:, -1].unsqueeze(0).detach().cpu().numpy().astype("int64")
-                    for i in range(int((len(outputs) - 1) / 2)):
-                        ort_input["past_key_values.{}.key".format(i)] = outputs[i*2+1]
-                        ort_input["past_key_values.{}.value".format(i)] = outputs[i*2+2]
-                    ort_input["attention_mask"] =  np.zeros([self.batch_size, ort_input["past_key_values.0.key"].shape[2]+1], dtype="int64")
-                    yield ort_input, last_ind.detach().cpu().numpy()
-        except StopIteration:
-            return
-
 class GPTQDataloader:
-    def __init__(self, model_path, batch_size=1, seqlen=2048, sub_folder="train"):
+    def __init__(self, model_input, batch_size=1, seqlen=2048, sub_folder="train", sampling_size=8):
         import random
         random.seed(0)
         self.seqlen = seqlen
 
         self.batch_size=batch_size
-        # self.traindata = load_dataset(args.dataset, split=sub_folder)
-        self.traindata = datasets.load_from_disk('/mnt/data1/mengni/pile')['train']
-        self.traindata = self.traindata.map(tokenize_function, batched=True)
-        self.traindata.set_format(type="torch", columns=["input_ids", "attention_mask"])
-        self.sess = None
-        if not model_path.endswith("decoder_model.onnx"):
-            self.sess = ort.InferenceSession(os.path.join(os.path.dirname(model_path), "decoder_model.onnx"))
+        traindata = datasets.load_dataset(args.dataset, split=sub_folder)
+        traindata = traindata.map(tokenize_function, batched=True)
+        traindata.set_format(type="torch", columns=["input_ids", "attention_mask"])
+        self.traindata = traindata.select(range(sampling_size))
+
+        session = ort.InferenceSession(model_input)
+        inputs_names = [input.name for input in session.get_inputs()]
+        self.key_value_input_names = [key for key in inputs_names if (".key" in key) or (".value" in key)]
+        self.use_cache = len(self.key_value_input_names) > 0
 
     def __iter__(self):
         try:
-            while True:
+            for _ in range(self.sampling_size):
```

Comment:
sampling_size is not a member of the class. You need to add it as a member.

Reply:
Fixed. Thanks for review.